### PR TITLE
Rekey V2

### PR DESCRIPTION
**Pull request by maxtaco**
- do everything within a singleton -- one per process. The struct is called a RekeyMaster
  - have to always have a UI around, even it it's not visible, since otherwise we can't send debug UI events
  - introduce a new flag that captures if the UI is visible or not.  I'm slightly worried that this flag
    will get unsynced and therefore we won't send UI clears as we ought to
- an (ambitious) testing strategy
  - introduce device provisioning and multiple services in systests
- contextify posts in some cases so that paper key provisioning works in systests
- unflakify flakey test: it was a race over reserve_sig in JSON blobs
Verified

### CODE REVIEW

**Comment by maxtaco** [comment]
@patrickxb can you help me review this?

CC: @keybase/react-hackers, I added a [new RPC](https://github.com/keybase/client/pull/3877/files#diff-b35df65f797420e7aa29ea330b920222R33) to the `RekeyUI` protocol.  It's for testing and debugging, so it's safe to ignore.

File: protocol/avdl/rekey_ui.avdl [hunk 1]
```diff
@@ -4,8 +4,31 @@ protocol rekeyUI {
...
+  record RekeyEvent {
+    RekeyEventType eventType;
+    int interruptType;
+  }
+
+  /**
+   RekeySendEvent sends updates as to what's going on in the rekey
+   thread. This is mainly useful in testing.
+   */
+  void rekeySendEvent(int sessionID, RekeyEvent event);
```
Comment (maxtaco) [inline protocol/avdl/rekey_ui.avdl]: CC: @keybase/react-hackers, here is the new RPC.

File: go/vendor/github.com/keybase/go-jsonw/jsonw.go [hunk 2]
```diff
@@ -351,6 +351,7 @@ func (rd *Wrapper) GetString() (ret string, err error) {
 	} else if b, ok := rd.dat.([]byte); ok {
 		ret = string(b)
 	} else {
+		//fmt.Printf("GetString [%p] %v\n", rd, rd.dat)
```
Comment (patrickxb) [inline go/vendor/github.com/keybase/go-jsonw/jsonw.go]: delete

File: go/vendor/github.com/keybase/go-jsonw/jsonw.go [hunk 3]
```diff
@@ -455,6 +456,7 @@ func (rd *Wrapper) AtKey(s string) *Wrapper {
 			ret.err = ret.NewError("no such key: %s", s)
 		}
 	}
+	//fmt.Printf("AtKey %s -> %v\n", s, ret)
```
Comment (patrickxb) [inline go/vendor/github.com/keybase/go-jsonw/jsonw.go]: delete

File: go/vendor/github.com/keybase/go-jsonw/jsonw.go [hunk 4]
```diff
@@ -531,6 +533,7 @@ func tryInt(bit string) (ret int, isInt bool) {
 }
 
 func (w *Wrapper) AtPath(path string) (ret *Wrapper) {
+	//fmt.Printf("AtPath %p %s\n", w, path)
```
Comment (patrickxb) [inline go/vendor/github.com/keybase/go-jsonw/jsonw.go]: delete

File: protocol/Makefile [hunk 5]
```diff
@@ -11,7 +11,7 @@ json/%.json: avdl/%.avdl
 	$(AVDL2JSON) -i $< -o $@~ && mv $@~ $@
 
 config:
-	npm i
+	# npm i
```
Comment (patrickxb) [inline protocol/Makefile]: put this back

**Comment by patrickxb** [comment]
LGTM!

File: protocol/avdl/rekey_ui.avdl (hunk 1 above)
Comment (chrisnojima) [inline protocol/avdl/rekey_ui.avdl]: adding a handler

**Pull request merged**